### PR TITLE
Client-side translations: gettext support

### PR DESCRIPTION
**Pull request by Ekdohibs**
Subsumes #9282.

----

Add gettext support for client-side translations. This allows using existing gettext tooling, and will allow using plurals in translated strings (`ngettext` and the like).

## To do

This PR is a Work in Progress.

- [x] Support MO files
- [x] Support plurals
- [ ] Testing
  + [ ] escape sequence handling, on different platforms (especially Windows) for the handling of surrogate pairs
  + [ ] Lots of PO files (to check that the parser works with all automatically-produced PO files)
  + [ ] Malformed files (check against possible attack by buffer overflow in the parsing code)
  + [ ] MO files (completely untested for now)
  + [ ] plurals (completely untested for now)
- [x] Add tooling for extraction of strings?

## How to test

For now, take a mod using client-side translations, and replace the .tr file with a .po one, with the following entries for each original translated string:
```
msgctx "<textdomain>"
msgid "<original string>"
msgstr "<translated string>"
```

### CODE REVIEW

**Comment by Ekdohibs** [comment]
Does anyone know why `ngettext` is not found in windows builds?

**Comment by nerzhul** [comment]
outside of the PR scope, can we push translations from server to client to load it ? it will be just very very nice for CSM mods or some other things we can push from server ie. nodes

**Comment by rubenwardy** [comment]
> outside of the PR scope, can we push translations from server to client to load it ? it will be just very very nice for CSM mods or some other things we can push from server ie. nodes

This already happens, client side translations are client side. Do you mean all possible translations rather than just the user language?

**Comment by Ekdohibs** [comment]
Translations are client-side, and the translations are pushed from the server to the client. However, the client receives indeed only the translations in its language, to save bandwidth.

**Comment by Pevernow** [comment]
I would like to merge this request.

This is important for many mobile phone users.

File: src/util/string.cpp [hunk 1]
```diff
@@ -144,6 +144,16 @@ std::string wide_to_utf8(const std::wstring &input)
 	return out;
 }
 
+void wide_add_codepoint(std::wstring &result, char32_t codepoint)
+{
+	if ((0xD800 <= codepoint && codepoint <= 0xDFFF) || (0x10FFFF < codepoint)) {
```
Comment (red-001) [inline src/util/string.cpp:149]: Couldn't the code for checking if the code point is valid be shared between the two implementations?

File: src/util/string.cpp [hunk 2]
```diff
@@ -858,10 +903,37 @@ void translate_all(const std::wstring &s, size_t &i,
...
+					if (L'0' <= c && c <= L'9') {
+						number = (10 * number + (unsigned long int)(c - L'0'));
+						if (number >= 2 * max) number = (number % max) + max;
+					} else {
+						// Invalid number
+						use_plural = false;
+						break;
+					}
+				}
+			}
```
Comment (red-001) [inline src/util/string.cpp:934]: Shouldn't this already be done in Lua? Seems like it could just convert the string to a number.

File: src/translation.cpp [hunk 3]
```diff
@@ -143,15 +220,335 @@ void Translations::loadTranslation(const std::string &data)
...
+		return;
+	} else if (!removeStringEnd(filename, poExtension).empty()) {
+		loadPoTranslation(data);
+		return;
+	} else if (!removeStringEnd(filename, moExtension).empty()) {
+		loadMoTranslation(data);
+		return;
+	} else {
+		errorstream << "loadTranslation called with invalid filename: \"" << filename << "\"" << std::endl;
+		return;
```
Comment (red-001) [inline src/translation.cpp:552]: The early returns don't change code flow and just confuse the reader.

File: src/translation.cpp [hunk 4]
```diff
@@ -143,15 +220,335 @@ void Translations::loadTranslation(const std::string &data)
...
+					c = c * 16 + v;
+				}
+				if (j == 0) {
+					errorstream << "Invalid escape sequence \\x, ignoring" << std::endl;
+					continue;
+				}
+				// If character fits in 16 bits and is not part of surrogate pair, insert it.
+				// Otherwise, silently drop it: this is valid since \x escape sequences with
+				// values above 0xff are implementation-defined
+				if ((c < 0xd800) || (0xe000 <= c && c <= 0xffff)) result.push_back(c);
```
Comment (red-001) [inline src/translation.cpp:286]: another single line if

File: src/translation.cpp [hunk 5]
```diff
@@ -33,12 +35,13 @@ Translations *g_client_translations = &client_translations;
...
 {
 	m_translations.clear();
+	m_plural_translations.clear();
 }
 
 const std::wstring &Translations::getTranslation(
 		const std::wstring &textdomain, const std::wstring &s)
 {
-	std::wstring key = textdomain + L"|" + s;
+	std::wstring key = textdomain + L"|"; key.append(s);
```
[the remark this answers is not in the document]
Comment (Ekdohibs) [reply src/translation.cpp:44]: Yes. Although maybe the compiler could optimize that? I'm not sure.

File: src/util/string.cpp (hunk 2 above)
Comment (Ekdohibs) [reply src/util/string.cpp:934]: Unfortunately not: Lua has to convert the number to a string to encode it inside the translation string. However, it is possible to create invalid translation strings without using `minetest.translate`, and these must not crash the code.

File: src/util/string.cpp (hunk 1 above)
Comment (Ekdohibs) [reply src/util/string.cpp:149]: Probably, I can change that if you want. However, since the code on WIN32 also compares the codepoint to 0x10000 in all cases, I thought it was more efficient to separate the condition.

File: src/translation.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (red-001) [reply src/translation.cpp:44]: I feel it might require too much knowledge of how strings work. Checked godbolt and it doesn't seem to. https://godbolt.org/z/8KrEWj

File: src/util/string.cpp (hunk 1 above)
Comment (red-001) [reply src/util/string.cpp:149]: I imagine the compiler would happily do that on it's own, but I suppose this is perfectly fine too. It's not like valid unicode codepoints will change any time soon.

File: src/util/string.cpp (hunk 2 above)
Comment (red-001) [reply src/util/string.cpp:934]: Oh my bad I thought the limit was just to avoid incorrect translations.

Comment (Ekdohibs) [reply src/util/string.cpp:934]: Ah, you meant the truncation to 6 digits should already be done by Lua? In that case yes, but I prefer to be conservative about the inputs received from the server. However, I can remove it if you prefer.

Comment (red-001) [reply src/util/string.cpp:934]: No I suppose it's good to make sure, could you make `max` a `constexpr`?

**Comment by rubenwardy** [comment]
I would really like to see this done. Using the gettext file format means that modders can use tools like weblate, and also port over translations from intllib more easily

**Comment by Ekdohibs** [comment]
I think this PR is mostly ready, it just needs to be tested. Will rebase.

**Comment by Ekdohibs** [comment]
Rebased!

File: src/translation.cpp [hunk 6]
```diff
@@ -143,14 +220,338 @@ void Translations::loadTranslation(const std::string &data)
...
+			continue;
+		}
+
+		const std::string original(cdata + original_off, original_len);
+		const std::string translated(cdata + translated_off, translated_len);
+
+		loadMoEntry(original, translated);
+	}
+	
+	warningstream << "mo files are unsupported for now" << std::endl;
```
Comment (sfan5) [inline src/translation.cpp:540]: but they _do_ work, right?

File: src/util/string.cpp [hunk 7]
```diff
@@ -164,6 +174,29 @@ std::string wide_to_utf8(const std::wstring &input)
...
+	codepoint -= 0x10000;
+	if (codepoint >= 0x100000) {
+		// original codepoint was above 0x10FFFF, so invalid
+		// replace with unicode replacement character
+		result.push_back(0xFFFD);
+		return;
+	}
+	result.push_back((wchar_t) ((codepoint >> 10) | 0xD800));
+	result.push_back((wchar_t) ((codepoint & 0x3FF) | 0xDC00));
+}
```
Comment (sfan5) [inline src/util/string.cpp:198]: The assumption here seems to be that `wchar_t` is 2 bytes only on Windows and 4 bytes on all other platforms.
I don't think this is a good idea.

File: src/util/string.cpp [hunk 8]
```diff
@@ -794,10 +839,37 @@ void translate_all(const std::wstring &s, size_t &i,
...
+				use_plural = true;
+				number = 0;
+				for (char c : parts[2]) {
+					if (L'0' <= c && c <= L'9') {
+						number = (10 * number + (unsigned long int)(c - L'0'));
+						if (number >= 2 * max) number = (number % max) + max;
+					} else {
+						// Invalid number
+						use_plural = false;
+						break;
```
Comment (sfan5) [inline src/util/string.cpp:867]: If builtin already does this, why do it again here?
If the server sends a number that's too big and it truncates incorrectly, that's not our problem.

File: src/translation.cpp [hunk 9]
```diff
@@ -143,14 +220,338 @@ void Translations::loadTranslation(const std::string &data)
...
+				i++;
+				if (i + 4 > str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char16_t c = 0;
+				bool ok = true;
+				for (size_t j = 0; j < 4; j++) {
+					unsigned char v;
+					if (str[i+j] <= 0xff && hex_digit_decode((char)str[i+j], v)) {
+						c = c * 16 + v;
```
Comment (sfan5) [inline src/translation.cpp:303]: I know it's pretty simple but do we really need to reimplement number parsing here?

If it's kept it should at least be refactored into a helper function (I suggest an inline lambda since it's only needed by this function).

File: src/translation.cpp [hunk 10]
```diff
@@ -143,14 +220,338 @@ void Translations::loadTranslation(const std::string &data)
...
+			}
+			case L'x': {
+				i++;
+				if (i >= str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char32_t c = 0;
+				size_t j = 0;
+				unsigned char v;
+				for (; i+j < str.length() && hex_digit_decode((char)str[i+j], v); j++) {
```
Comment (sfan5) [inline src/translation.cpp:277]: Do I see it correctly that `\x` takes an arbitrary number of hex characters?
Shouldn't that be exactly two?

File: src/server.cpp [hunk 11]
```diff
@@ -2527,19 +2527,38 @@ void Server::sendMediaAnnouncement(session_t peer_id, const std::string &lang_co
...
+	std::string translation_formats[3] = { ".tr", ".po", ".mo" };
+	std::string lang_suffixes[3];
+	for (size_t i = 0; i < 3; i++) {
+		lang_suffixes[i].append(".").append(lang_code).append(translation_formats[i]);
+	}
 	for (const auto &i : m_media) {
-		if (str_ends_with(i.first, ".tr") && !str_ends_with(i.first, lang_suffix))
+		bool ok = true;
+		for (size_t j = 0; j < 3; j++) {
+			if (str_ends_with(i.first, translation_formats[j]) && !str_ends_with(i.first, lang_suffixes[j])) {
```
Comment (sfan5) [inline src/server.cpp:2538]: Could use `removeStringEnd` instead of `str_ends_with(i.first, translation_formats[j])` here

File: src/translation.h [hunk 12]
```diff
@@ -30,11 +32,26 @@ extern Translations *g_client_translations;
...
 	std::unordered_map<std::wstring, std::wstring> m_translations;
+	std::unordered_map<std::wstring, std::vector<std::wstring>> m_plural_translations;
+
+	void addTranslation(const std::wstring &textdomain, const std::wstring &original,
+			const std::wstring &translated);
+	void addPluralTranslation(const std::wstring &textdomain,
+			const std::wstring &original,
+			std::vector<std::wstring> &translated);
+	std::wstring unescapeC(const std::wstring &str);
+	void loadPoEntry(const std::map<std::wstring, std::wstring> &entry);
```
Comment (rubenwardy) [inline src/translation.h:52]: This is optional: but I feel like there are 3 concerns here mixed into one class. 1) Querying translations, 2) parsing .po, and 3) parsing.tr.

It would be nicer if the parsers were moved into a different class, you could then make a unit test which reads a .po file with various edge cases inside

File: src/translation.cpp [hunk 13]
```diff
@@ -143,14 +220,338 @@ void Translations::loadTranslation(const std::string &data)
...
+	
+	warningstream << "mo files are unsupported for now" << std::endl;
+	return;
+}
+
+void Translations::loadTranslation(const std::string &filename, const std::string &data) {
+	const char *trExtension[] = { ".tr", NULL };
+	const char *poExtension[] = { ".po", NULL };
+	const char *moExtension[] = { ".mo", NULL };
+	if (!removeStringEnd(filename, trExtension).empty()) {
```
Comment (rubenwardy) [inline src/translation.cpp:548]: I suggest making an `endsWith` function to be used like so:

```c++
if (str_ends_with(filename, ".tr")) {
```

**Review by rubenwardy — COMMENTED** [review]
Haven't reviewed .po parsing yet

**Comment by rubenwardy** [comment]
Bump
